### PR TITLE
Separate MessageHeader into Packet/Payload/Mac

### LINKED ISSUE #2716: Split message header into separate components (single encode/decode methods)
#### Problem
MessageHeader grew organically and now has Encode/Decode, EncodeEncryptedHeader/DecodeEncryptedHeader and EncodeMacTag/DecodeMacTag

This makes it hard to follow which items are actually in each part of the message.

 #### Proposed Solution
Make messageheader as a container class (potentially renamed, as it is not a single header anymore) and split out each of non encrypted/encrypted/mac into separate classes that are independent.

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Header fields were assumed to be a single structure, even though the sources are different (some encrypted, some not, different usages)

 #### Summary of Changes
Split headers into packet/payload and mac. This allows for more scoped code (e.g. for mac) and more targeted processing.
It should also make it easier for reliable messaging to determine what pieces of information are available where.

fixes #2716

### CODE REVIEW

**Review by rwalker-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@hawk248 @BroderickCarlin @saurabhst @jelderton ?

**Review by BroderickCarlin — APPROVED** [review]
